core/executor/src/wasm_executor.rs

```diff
@@ -657,7 +661,69 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
 			5
 		})
 	},
-	ext_sr25519_verify(msg_data: *const u8, msg_len: u32, sig_data: *const u8, pubkey_data: *const u8) -> u32 => {
+	ext_ed25519_generate(id_data: *const u8, seed: *const u8, seed_len: u32, out: *mut u8) => {
+		let mut id = [0u8; 4];
+		this.memory.get_into(id_data, &mut id[..])
+			.map_err(|_| "Invalid attempt to get id in ext_ed25519_generate")?;
+		let key_type = KeyTypeId(id);
+
+		let seed = if seed_len == 0 {
+			None
+		} else {
+			Some(
+				this.memory.get(seed, seed_len as usize)
+					.map_err(|_| "Invalid attempt to get seed in ext_ed25519_generate")?
+			)
+		};
+
+		let seed = seed.as_ref()
+			.map(|seed|
+				std::str::from_utf8(&seed)
+					.map_err(|_| "Seed not a valid utf8 string in ext_sr25119_generate")
+			).transpose()?;
+
+		let pubkey = runtime_io::ed25519_generate(key_type, seed);
+
+		this.memory.set(out, pubkey.as_ref())
+			.map_err(|_| "Invalid attempt to set out in ext_ed25519_generate".into())
+	},
+	ext_ed25519_sign(
+		id_data: *const u8,
+		pubkey_data: *const u8,
+		msg_data: *const u8,
+		msg_len: u32,
+		out: *mut u8,
+	) -> u32 => {
+		let mut id = [0u8; 4];
+		this.memory.get_into(id_data, &mut id[..])
+			.map_err(|_| "Invalid attempt to get id in ext_ed25519_sign")?;
+		let key_type = KeyTypeId(id);
+
+		let mut pubkey = [0u8; 32];
+		this.memory.get_into(pubkey_data, &mut pubkey[..])
+			.map_err(|_| "Invalid attempt to get pubkey in ext_ed25519_sign")?;
+
+		let msg = this.memory.get(msg_data, msg_len as usize)
+			.map_err(|_| "Invalid attempt to get message in ext_ed25519_sign")?;
+
+		let signature = runtime_io::ed25519_sign(key_type, &pubkey, &msg);
```

Comment:
this seems as though it could easily break consensus

Reply:
i think it can only be called from the offchain worker. if not, then it needs to change.

Reply:
We have at least 3 different "capabilities" for runtime calls that affect the set of callable APIs:
1. Offchain extensions available (offchain workers)
2. Keystore available (offchain workers, rotate session, equivocation report generation)
3. Regular runtime calls

So regular, consensus-critical execution path will neither have keystore nor offchain extensions set so those APIs will be disabled (or will just always return an error).

The offchain and keystore CryptoApis are now a bit duplicated, but afaik @bkchr is working on unification (@bkchr let me know if you need any help).

Reply:
Yeah, the `keystore` argument is optional and will be `None` for block production/block importing. Thus, these apis will not work.

After fixing the tests again, I will start to clean this stuff up.